system/system_error_monitor/src/system_error_monitor_core.cpp

```diff
@@ -26,8 +26,27 @@
 
 #include <fmt/format.h>
 
+#define DEFINE_TIER4_RCLCPP_LOG(LEVEL)                                                   \
+  void TIER4_RCLCPP_##LEVEL##_THROTTLE(                                                  \
+    const std::string & logger_name, rclcpp::Clock clock, const double duration,         \
+    const std::string & message)                                                         \
+  {                                                                                      \
+    static std::unordered_map<std::string, rclcpp::Time> last_output_time;               \
+    if (last_output_time.count(logger_name) != 0) {                                      \
+      const auto time_from_last_output = clock.now() - last_output_time.at(logger_name); \
+      if (time_from_last_output.seconds() * 1000.0 < duration) {                         \
+        return;                                                                          \
+      }                                                                                  \
+    }                                                                                    \
+    last_output_time[logger_name] = clock.now();                                         \
+    RCLCPP_##LEVEL(rclcpp::get_logger(logger_name), message.c_str());                    \
+  }
+
```

Comment:
It's better to replace this with `RCUTILS_LOG_COND_NAMED` or `template <int log_level>`.

Reply:
`RCUTILS_LOG_COND_NAMED` requires `RCUTILS_LOG_CONDITION_ONCE_BEFORE`, `RCUTILS_LOG_CONDITION_ONCE_AFTER`, ... which makes the readability less, so I used if else.

I'm not sure about `if constexpr`, but is this code what you intended to do?

Reply:
> RCUTILS_LOG_COND_NAMED requires RCUTILS_LOG_CONDITION_ONCE_BEFORE, RCUTILS_LOG_CONDITION_ONCE_AFTER, ... which makes the readability less

Could you explain this in a bit more detail?

> I'm not sure about if constexpr, but is this code what you intended to do?

Yes, almost perfect.
But it's better to use `enum class` instead of `int`.
Or if there are some macros like `RCLCPP_LOG_LEVEL_DEBUG`, you can use it instead.

Reply:
Arguments of `RCUTILS_LOG_COND_NAMED` is like this.
![image](https://user-images.githubusercontent.com/20228327/158014701-819620ca-8915-44f1-8341-9fdf651af439.png)
, and this is an example of macros for before and after.
![image](https://user-images.githubusercontent.com/20228327/158014698-93760124-b69c-40ec-a171-1662ba69707a.png)

Reply:
>Or if there are some macros like RCLCPP_LOG_LEVEL_DEBUG, you can use it instead.

There is defined values like this, which I don't think good to use in the application side, so I didn't use them.
![image](https://user-images.githubusercontent.com/20228327/158014887-a26cc247-03e4-4dd3-afe4-a62eddc464af.png)

Reply:
https://github.com/autowarefoundation/autoware.universe/pull/510/commits/a767e1aa6e03e5eb53b326ca27b6221265855927
DONE

Reply:
> , and this is an example of macros for before and after.

Hmm? Why `RCUTILS_LOG_CONDITION_ONCE_BEFORE` and `RCUTILS_LOG_CONDITION_ONCE_AFTER` are required?

Reply:
You mean, it's ok just to put empty for BEFORE and AFTER? If so, it's right. Anyway we have to use flag like `RCUTILS_LOG_MIN_SEVERITY_DEBUG`, which I don't think good as I said.

Reply:
> Do you mean it's ok just to put empty for BEFORE and AFTER? If so, it's right. 

Yes, I think it's okay.

> Anyway we have to use flag like RCUTILS_LOG_MIN_SEVERITY_DEBUG, which I don't think good as I said.

I think it's acceptable because it's documented.
https://docs.ros2.org/bouncy/api/rclcpp/logging_8hpp.html#a28cf59074387c81ef082f3e774481e95

Reply:
Just got it.